### ISSUE TITLE
Custom "Authorization" headers not allowed?

### ISSUE THREAD

**Issue by Stegosource** [smallwins/begin-issues#19]
**Expected behavior**
I should be able to send a custom "Authorization" header with my request, and read it on the backend.

**Description of bug**
The backend fails whenever I send a fetch request with the "Authorization" header present. Not able to access headers or any sort of error message. Client console receives:
```
Access to fetch at 'http://localhost:3333/login' from origin 'http://localhost:8080' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: No 'Access-Control-Allow-Origin' header is present on the requested resource. If an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled.
```

**Steps to reproduce**
Detailed steps to reproduce the behavior:
1. Create backend function:
```js
exports.handler = async (req) => {
    return {
      statusCode: 200,
      headers: {
        "Access-Control-Allow-Origin": "*",
      },
      body: JSON.stringify({
        hello: "world!",
      }),
    };
};
```
2. Send fetch request
```js
fetch("http://localhost:3333/handler", {
  "headers": {
    "Authorization": "Bearer yo"
  },
})
```

**Additional context or notes**
I also tested with a POST request, `mode: 'no-cors'`, and `credentials: 'include'`

**Comment by ryanblock** [comment]
Will have to take a peek in the morning, but it’s CORS, so you are likely aware how fiddly that can often be even in the best circumstances. In the mean time, can you run the entire app through sandbox?

**Comment by Stegosource** [comment]
Yeah, no worries. Take your time. 

Im sorry, but I dont know what you mean. I assume you mean `@architect/sandbox`, and if that's the case then I can only say that I've been deving with the command `cross-env NODE_ENV=testing npx sandbox`

**Comment by ryanblock** [comment]
In your error above, your request originated from localhost:8080, not :3333, so I assumed you had some second web server running that was maybe messing up CORS. Do you know where the :8080 came from?

**Comment by Stegosource** [comment]
Yeah, it's just a Vue CLI dev server. I dont think it should have any impact on CORS. If that were the case, then all my other requests be blocked, right? But I'm able to make every other GET or POST request as long as I don't add the custom Auth headers.

The following works:
```js
fetch("http://localhost:3333/login", {"headers":{"content-type":"application/x-www-form-urlencoded;charset=UTF-8"},"body":"login=temp&password=12345678","method":"POST"});
```

The following doesn't:
```js
fetch("http://localhost:3333/login", {"headers":{"content-type":"application/x-www-form-urlencoded;charset=UTF-8", "Authorization": "Bearer yo"},"body":"login=temp&password=12345678","method":"POST"});
```

The only difference is the Authorization header :/

**Comment by ryanblock** [comment]
Yeah, you're right, it doesn't have anything to do with Vue. So here's what I'm seeing:

With added presence of the Authorization header (and this applies to other headers, too), the fetch automatically API adds a CORS preflight request (`OPTIONS`).

This preflight request isn't being handled, which fails the request at the client (which then doesn't send your request at all, including your header).

You can test this by using any other client to send a regular HTTP request:
`curl -i -H "Authorization: foo" http://localhost:3333` → `console.log(req.headers.Authorization) // foo`

Further, here's a test I wrote a little while back to ensure receipt and proper AWS-like handling of the `Authorization` header: https://github.com/architect/sandbox/blob/master/test/unit/src/http/invoke-http/_header-fmt-test.js#L18-L26

tl;dr - this is a CORS issue related to handling fetch / CORS, and is unrelated to Sandbox / Begin. I'll keep poking at it but I have to jump on a call, so I wanted to at least give you a bug vector to work with in the mean time!

**Comment by Stegosource** [comment]
Yeah, I kind of got this far along as well. I'm not sure how to handle the OPTIONS preflight (or if I can manually), but I'm glad that I'm not crazy. 

Maybe with your other comments about using Express I can find a simple workaround with the `cors` library.

**Comment by ryanblock** [comment]
You're def not crazy! Although CORS is definitely crazy-making – we don't jokingly refer to it as "cross origin ritual sacrifice" for nothing.

You can definitely handle it manually (`if (req.httpMethod === 'OPTIONS')`, etc.), although to your point you may prefer to delegate that out to another tool.

Will close for now but please let me know if I can do anything else to aid here!

**Comment by Stegosource** [comment]
Well, I still dont understand how this works. Adding a `/src/http/options-handler/index.js` doesnt work because adding `options /handler` to `.arc`  throws
```
× Error: Architect schema validation error
  @http invalid route: "options /handler"
```
And trying with `if (req.httpMethod === 'OPTIONS')` on the POST route handler doesn't work because the code execution never gets into the POST handler.

The body of that POST handler is just:
```js
exports.handler = async (req) => {
    return {
      statusCode: 200,
      headers: {
        "Access-Control-Allow-Origin": "*",
      },
      body: JSON.stringify({
        hello: "world!",
      }),
    };
};
```

**Comment by ryanblock** [comment]
Sorry, I should have added some more information to that. The preflight request should hit the route you requested, so if you're calling `"http://localhost:3333/login`, `src/http/get-login` should be receiving the preflight.

I'm also gonna see if I can whip together an example today.

**Comment by Stegosource** [comment]
When you say that it should, then why doesn't the most basic handler work?

This should always return, right? 
```
exports.handler = async (req) => {
    return {
      statusCode: 200,
      headers: {
        "Access-Control-Allow-Origin": "*",
      },
      body: JSON.stringify({
        hello: "world!",
      }),
    };
};
```
But in my testing, this is not the case if the "Authorization" header is present. The OPTIONS request fails, and Im not sure how to handle it so it doesn't fail :/

**Comment by Stegosource** [comment]
I tested this again with Express and CORS and it still does not work. Here's my example:

```js
const arc = require("@architect/functions");
const app = require("express")();

app.use(require("cors")());

app.post("/test", (req, res) => {
  console.log(req);
  res.send("very cool");
});

exports.handler = arc.http.express(app);
```

Frontend (tried a POST request, but the OPTIONS request failed first)
```js
fetch("http://localhost:3333/test", {
  credentials: "omit",
  headers: {
    accept: "*/*",
    "accept-language": "en-US,en;q=0.9",
    "cache-control": "no-cache",
    pragma: "no-cache",
    "sec-fetch-dest": "empty",
    "sec-fetch-mode": "cors",
    "sec-fetch-site": "same-site",
  },
  referrer: "http://localhost:8080/dashboard/",
  referrerPolicy: "no-referrer-when-downgrade",
  body: null,
  method: "OPTIONS",
  mode: "cors",
});
```

Browser error:
```
OPTIONS http://localhost:3333/cards 404 (Not Found)

Access to fetch at 'http://localhost:3333/test' from origin 'http://localhost:8080' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: No 'Access-Control-Allow-Origin' header is present on the requested resource. If an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled.
```

Unfortunately there doesnt seem to be any way to work with the OPTIONS request.

**Comment by ryanblock** [comment]
Hey, I haven't forgotten about this! I've got it working locally, but need to do some more testing and verification to deployed infra. Please stand by!

**Comment by ryanblock** [comment]
Ok! So, we do have some work to improve how we're doing `options` preflights, but this code should work.

Let's start with your original requester:
```
fetch("http://localhost:3333/login", {
  "headers": {
    "content-type": "application/x-www-form-urlencoded;charset=UTF-8", 
    "Authorization": "Bearer yo"
  },
  "body": "login=temp&password=12345678",
  "method":"POST"
});
```

Add this to your `src/http/get-index/index.js` `handler` function right up top:
```
if (req.httpMethod === 'OPTIONS') {
    return {
      statusCode: 200,
      headers: {
        'Access-Control-Allow-Methods': 'GET, OPTIONS',
        'Access-Control-Allow-Origin': '*',
        'Access-Control-Allow-Headers': 'Authorization, authorization'
      }
    }
  }
```

And as long as `src/http/get-login/index.js` has `'Access-Control-Allow-Origin': '*'` in its response headers, you should be good to go.

A future fix we will issue will allow you to define `options` in `@http`, e.g.:
```
@http
get /login
options /login
```

**Comment by Stegosource** [comment]
Well, I'm not sure if I got this right yet. I just tested again and it still does not work. My entire handler file looks like this now:

```js

exports.handler = async (req) => {
  if (req.httpMethod === "OPTIONS") {
    return {
      statusCode: 200,
      headers: {
        "Access-Control-Allow-Methods": "GET, POST, OPTIONS",
        "Access-Control-Allow-Origin": "*",
        "Access-Control-Allow-Headers": "Authorization, authorization",
      },
    };
  }

  return {
    statusCode: 200,
    headers: {
      "Access-Control-Allow-Methods": "GET, POST, OPTIONS",
      "Access-Control-Allow-Origin": "*",
      "Access-Control-Allow-Headers": "Authorization, authorization",
    },
    body: JSON.stringify({
      hello: "world!",
    }),
  };
};
```

Thats all. 

And the entire fetch (copied from network request in dev tools):
```js
fetch("http://localhost:3333/login", {
  credentials: "omit",
  headers: {
    accept: "*/*",
    "accept-language": "en-US,en;q=0.9",
    "cache-control": "no-cache",
    pragma: "no-cache",
    "sec-fetch-dest": "empty",
    "sec-fetch-mode": "cors",
    "sec-fetch-site": "same-site",
  },
  referrer: "http://localhost:8080/login",
  referrerPolicy: "no-referrer-when-downgrade",
  body: null,
  method: "OPTIONS",
  mode: "cors",
});
```

I just upgraded to the latest sandbox as well, so I dont think it's that.

**Comment by ryanblock** [comment]
`my entire handler file looks like this now` which handler? Above I specified that options conditional needs to be in `src/http/get-index/index.js`.

**Comment by Stegosource** [comment]
Ahh shit. I'm sorry. I thought it was still src/http/get-**index**/index.js. I didn't have a default `get /` defined in my `.arc` file, but now that I do, and added that header snippet, it is working.

Thanks so much, and I'm very sorry for all the hassle on this. Can I buy you a beer?

**Comment by ryanblock** [comment]
No problem. I was almost done with that example repo, too! 🍻

Also, please see: https://github.com/architect/architect/issues/760 (thoughts?)

**Comment by brianleroux** [comment]
eh everyone, it occurred to me this morning that using `Authorization` clientside mean the token being used is accessible to clientside js. this is generally considered to be a bad idea. it isn't insecure, per se, but it is definitely less secure than using HTTPOnly cookies for tokens. architect by default ships with HTTPOnly cookies/session support so end developers can have secure auth without resorting to clientside hacks (on the same origin). LMK if this stuff needs clarification or help.

**Comment by Stegosource** [comment]
@brianleroux thanks for chiming in. Im aware that arc provides httpOnly tokens, but Im not as familiar with using them, and I know they have their own pros/cons. Each fetch has to define the `credentials: 'include'` option which then requires the backend allowed origin headers to be more specifically defined. That can be a good thing, but also requires more work and thoughts around the http requests.

Regarding security, Im only storing an Auth token in memory, and I was going to use the httpOnly token for refreshing the Auth token, following advice from this article https://hasura.io/blog/best-practices-of-using-jwt-with-graphql/

Would love to hear more of your input on it, and maybe the HTTP only tokens alone are the way to go, but I'm also trying to build within the context of my knowledge. With Ryan's solution, everything seems to work fine.
